### PR TITLE
cocoa

### PR DESCRIPTION
**Pull request by stal888**
Kita. It is ready to merge at this point but I'll add more stuff depending on how long you let this sit.
The binary is here. https://zodiaclabs.org/storage/c1/uTox.dmg

### What works
- [x] Everything that was in xlib uTox
- [x] Rich input (tested with japanese and unicode picker, mmv for other things)
- [x] Video chat
- [x] Desktop streaming
- [x] Notifications (10.8+)
- [x] Inline screenshot
- [x] Sandbox-readiness
- [x] ~~10.10 blur styles~~ Blocked by extended colour support, which needs this PR merged first.
- [x] More keyboard shortcuts
- [x] ~~dock badge~~ There will just be a dock icon like on windows
- [x] Icon

### CODE REVIEW

**Comment by stal888** [comment]
@irungentoo @GrayHatter I'm cutting it off here because all the core functionality is done and [I need this] to be merged by 23 May.

**Comment by stal888** [comment]
thank you my chinese cartoon friend

File: cocoa/drawing.m [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+}
+
+void setfont(int id) {
+    global_text_state._use_font = fonts[id];
+}
+
+uint32_t setcolor(uint32_t color) {
+    uint32_t ret = global_text_state._use_font_color;
+
+    if (global_text_state._use_font_color_ref)
```
Comment (GrayHatter) [inline cocoa/drawing.m]: utox is adopting toxcore style, this needs curly braces.

File: cocoa/drawing.m [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+    CGFloat comp[4] = {r, g, b, 1.0};
+    global_text_state._use_font_color_ref = CGColorCreate(cs, comp);
+    CGColorSpaceRelease(cs);
+
+    global_text_state._use_font_color = color;
+    return ret;
+}
+
+void reload_fonts(void) {
+    for (int i = 0; i < sizeof(fonts) / sizeof(CTFontRef); ++i)
```
Comment (GrayHatter) [inline cocoa/drawing.m]: this too, you probably want to grep regex for `).*\n`

and there's a few above this as well

File: cocoa/grabdesktop.m [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+- (void)keyDown:(NSEvent *)theEvent {
+    //NSLog(@"%@", theEvent);
+    switch (theEvent.keyCode) {
+        case kVK_Escape:
+            stardust_display_capping_done(self.isVideo, 1, self.window);
+            break;
+        default:
+            break;
```
Comment (GrayHatter) [inline cocoa/grabdesktop.m]: braces

File: cocoa/interaction.m [hunk 4]
```diff
@@ -0,0 +1,890 @@
...
+static inline void select_right_to_char(char_t c) {
+    EDIT *edit = edit_get_active();
+    int loc = edit_getcursorpos(), end = loc + edit_selection(edit, NULL, 0);
+
+    if (end > edit->length)
+        return;
+
+    if (edit->data[end] == c) {
+        end += 1;
+    }
```
Comment (GrayHatter) [inline cocoa/interaction.m]: braces, no braces, braces, no braces.... WTF dood?

File: cocoa/interaction.m [hunk 5]
```diff
@@ -0,0 +1,890 @@
...
+
+        CFMutableDataRef dat = CFDataCreateMutable(kCFAllocatorDefault, 0);
+        CGImageDestinationRef dest = CGImageDestinationCreateWithData(dat, kUTTypePNG, 1, NULL);
+        CGImageDestinationAddImage(dest, img, NULL);
+        CGImageDestinationFinalize(dest);
+        CFRelease(dest);
+
+        size_t size = CFDataGetLength(dat);
+        uint8_t *owned_ptr = malloc(size);
+        memcpy(owned_ptr, CFDataGetBytePtr(dat), size);
```
Comment (GrayHatter) [inline cocoa/interaction.m]: if NULL undefined behaviour... wanna wrap this with an error just incase malloc fails?

File: edit.c [hunk 6]
```diff
@@ -5,14 +5,18 @@ static EDIT *active_edit;
 static struct
 {
     STRING_IDX start, length;
-    STRING_IDX p1, p2, pm;
-    //TODO: pm field doesn't seem to be used. Remove?
+    STRING_IDX p1, p2;
```
Comment (GrayHatter) [inline edit.c]: can you rename this a bit more descriptively?

**Comment by GrayHatter** [comment]
github lie's it merges fine with git

**Comment by GrayHatter** [comment]
Don't have OSx so I cant build or test this. But @tsudoko says it works and I read through the code, so as soon as someone else say it works, and confirms it will compile. This is ready to merge.

File: edit.c (hunk 6 above)
Comment (stal888) [inline edit.c]: Those fields were there in original uTox. I'll rename the mark fields though.

**Comment by irungentoo** [comment]
Can you remove the pic from this PR?

**Comment by stal888** [comment]
in the frozen disk image?

**Comment by irungentoo** [comment]
yes, that one. 

Can't have any unfree stuff in the repo.

**Comment by stal888** [comment]
- [x] Add braces 
- [x] Remove non free image
- [x] Fix build

**Comment by stal888** [comment]
Cross your fingers

**Comment by stal888** [comment]
@irungentoo I've removed the bad file.

**Pull request merged**